File: apps/expo/modules/streamer/ios/StreamerServer.swift
import Foundation
import ReadiumGCDWebServer
import ZIPFoundation
import os.log

/// Errors that can occur during streaming operations
enum StreamerError: Error {
    case serverStartFailed(String)
    case archiveNotFound(String)
    case archiveOpenFailed(String, Error)
    case pageNotFound(Int)
    case pageExtractionFailed(Int, Error)
    case invalidArchiveFormat(String)
    case cacheDirectoryCreationFailed(Error)
}

/// Configuration for a book being streamed
struct BookConfig {
    let bookId: String
    let filePath: String
    let cacheDir: String

    /// Cached list of image files in the archive (naturally sorted)
    let imageFiles: [Entry]
    
    init(bookId: String, filePath: String, cacheDir: String) throws {
        self.bookId = bookId
        self.filePath = filePath
        self.cacheDir = cacheDir
        
        let archive = try ZipArchive(path: filePath)
        self.imageFiles = archive.getImageFiles()
    }
}

/// Manages the HTTP server for streaming pages from ZIP/CBZ archives
class StreamerServer {
    static let shared = StreamerServer()

    private let logger = Logger(subsystem: "com.stump.streamer", category: "server")

    private var webServer: ReadiumGCDWebServer?

    /// A cache of registered books (bookId -> BookConfig)
    private var books: [String: BookConfig] = [:]

    // TODO: Explore if better options in Swift
    /// Lock for thread-safe access to books
    private let booksLock = NSLock()

    /// Dispatch queue for background ZIP operations
    private let extractionQueue = DispatchQueue(label: "com.stump.streamer.extraction", qos: .userInitiated, attributes: .concurrent)

    // Note: Apparently 0 means auto-assign which is best I think
    private var port: UInt = 0

    /// Whether the server is currently running
    var isRunning: Bool {
        webServer?.isRunning ?? false
    }

    private init() {}

    // MARK: - Server Lifecycle

    /// Start the HTTP server if not already running
    /// - Parameter port: Port to use (0 for auto-assign)
    /// - Returns: The actual port the server is running on
    /// - Throws: StreamerError if server fails to start
    func startServer(port: UInt = 0) throws -> UInt {
        if isRunning {
            return self.port
        }

        let server = ReadiumGCDWebServer()

        // GET /books/{bookId}/pages/{page}
        server.addHandler(
            forMethod: "GET",
            pathRegex: "^/books/([^/]+)/pages/(\\d+)$",
            request: ReadiumGCDWebServerRequest.self
        ) { [weak self] request, completionBlock in
            guard let self = self else {
                completionBlock(ReadiumGCDWebServerErrorResponse(statusCode: 500))
                return
            }

            self.handlePageRequest(request: request, completionBlock: completionBlock)
        }

        let options: [String: Any] = [
            ReadiumGCDWebServerOption_Port: port,
            ReadiumGCDWebServerOption_BindToLocalhost: true,
            ReadiumGCDWebServerOption_AutomaticallySuspendInBackground: false
        ]

        do {
            try server.start(options: options)
            self.webServer = server
            self.port = server.port
            logger.info("Started on port \(server.port)")
            return server.port
        } catch {
            logger.error("Failed to start server: \(error.localizedDescription)")
            throw StreamerError.serverStartFailed("Failed to start server: \(error.localizedDescription)")
        }
    }

    /// Stop the HTTP server
    func stopServer() {
        webServer?.stop()
        webServer = nil
        port = 0
        logger.info("Stopped")
    }

    // MARK: - Book Management

    /// Register a book for streaming
    /// - Parameters:
    ///   - bookId: The book ID
    ///   - filePath: Path to the archive file
    ///   - cacheDir: Directory to cache extracted pages
    func registerBook(bookId: String, filePath: String, cacheDir: String) throws {
        booksLock.lock()
        defer { booksLock.unlock() }

        guard FileManager.default.fileExists(atPath: filePath) else {
            throw StreamerError.archiveNotFound(filePath)
        }

        try createCacheDirectoryIfNeeded(at: cacheDir)

        let config = try BookConfig(bookId: bookId, filePath: filePath, cacheDir: cacheDir)
        
        books[bookId] = config
        logger.debug("Registered book \(bookId) with \(config.imageFiles.count) pages")
    }

    /// Unregister a book and optionally clean up its cache
    /// - Parameters:
    ///   - bookId: The book ID to unregister
    ///   - deleteCache: Whether to delete cached pages
    func unregisterBook(bookId: String, deleteCache: Bool = false) {
        booksLock.lock()
        defer { booksLock.unlock() }

        guard let config = books[bookId] else { return }

        if deleteCache {
            try? FileManager.default.removeItem(atPath: config.cacheDir)
            logger.debug("Deleted cache for book \(bookId)")
        }

        books.removeValue(forKey: bookId)
        logger.debug("Unregistered book \(bookId)")
    }

    /// Get the URL for a specific page
    /// - Parameters:
    ///   - bookId: The book ID
    ///   - page: The page number (1-indexed)
    /// - Returns: The URL to access the page
    func getPageURL(bookId: String, page: Int) -> String? {
        guard isRunning else { return nil }
        return "http://localhost:\(port)/books/\(bookId)/pages/\(page)"
    }

    /// Get the total number of pages for a book
    /// - Parameter bookId: The book ID
    /// - Returns: The number of pages, or nil if book not found
    func getPageCount(bookId: String) -> Int? {
        booksLock.lock()
        defer { booksLock.unlock() }
        
        return books[bookId]?.imageFiles.count
    }

    // MARK: - Request Handling

    /// Handle a page request
    private func handlePageRequest(request: ReadiumGCDWebServerRequest, completionBlock: @escaping ReadiumGCDWebServerCompletionBlock) {
        let path = request.path.components(separatedBy: "/").filter({ !$0.isEmpty })
        
        guard path.count >= 4,
              path[0] == "books",
              path[2] == "pages",
              let pageNumber = Int(path[3]) else {
            completionBlock(ReadiumGCDWebServerErrorResponse(statusCode: 400))
            return
        }

        let bookId = path[1]

        booksLock.lock()
        guard let config = books[bookId] else {
            booksLock.unlock()
            logger.warning("Book not found: \(bookId)")
            completionBlock(ReadiumGCDWebServerErrorResponse(statusCode: 404))
            return
        }
        booksLock.unlock()

        if let cachedPagePath = getCachedPagePath(config: config, page: pageNumber),
           FileManager.default.fileExists(atPath: cachedPagePath) {
            serveCachedPage(path: cachedPagePath, completionBlock: completionBlock)
            return
        }

        // Cache miss so extract
        extractionQueue.async {
            do {
                try self.extractAndServePage(config: config, page: pageNumber, completionBlock: completionBlock)
            } catch {
                self.logger.error("Failed to extract page \(pageNumber) from \(bookId): \(error.localizedDescription)")
                completionBlock(ReadiumGCDWebServerErrorResponse(statusCode: 500))
            }
        }
    }

    // MARK: - Page Extraction

    /// Extract a page from the archive and serve it
    private func extractAndServePage(config: BookConfig, page: Int, completionBlock: @escaping ReadiumGCDWebServerCompletionBlock) throws {
        let archive = try ZipArchive(path: config.filePath)

        // Use cached image files list
        let imageFiles = config.imageFiles

        guard page > 0 && page <= imageFiles.count else {
            throw StreamerError.pageNotFound(page)
        }

        // Note: page is 1-indexed
        let entry = imageFiles[page - 1]

        let data = try archive.extractEntry(entry)

        // TODO: This might be too aggressive? Maybe just serve the data if we can't write to cacheDir?
        guard let cachedPath = getCachedPagePath(config: config, page: page, originalFilename: entry.path) else {
            throw StreamerError.invalidArchiveFormat("Could not determine file extension for page \(page)")
        }
        
        try data.write(to: URL(fileURLWithPath: cachedPath))

        let contentType = getContentType(for: entry.path)
        let response = ReadiumGCDWebServerDataResponse(data: data, contentType: contentType)
        DispatchQueue.main.async {
            completionBlock(response)
        }

        logger.debug("Extracted and served page \(page) for book \(config.bookId)")
    }

    /// Serve a page from cache
    private func serveCachedPage(path: String, completionBlock: @escaping ReadiumGCDWebServerCompletionBlock) {
        do {
            let data = try Data(contentsOf: URL(fileURLWithPath: path))
            let contentType = getContentType(for: path)
            let response = ReadiumGCDWebServerDataResponse(data: data, contentType: contentType)
            completionBlock(response)
            logger.debug("Served cached page at \(path)")
        } catch {
            logger.error("Failed to read cached page: \(error.localizedDescription)")
            completionBlock(ReadiumGCDWebServerErrorResponse(statusCode: 500))
        }
    }

    // MARK: - Helper Methods

    /// Create cache directory if it doesn't exist
    private func createCacheDirectoryIfNeeded(at path: String) throws {
        let url = URL(fileURLWithPath: path)
        if !FileManager.default.fileExists(atPath: path) {
            do {
                try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
            } catch {
                throw StreamerError.cacheDirectoryCreationFailed(error)
            }
        }
    }

    /// Get the cached page path with correct extension
    private func getCachedPagePath(config: BookConfig, page: Int, originalFilename: String? = nil) -> String? {
        if let filename = originalFilename {
            let ext = (filename as NSString).pathExtension
            guard !ext.isEmpty else { return nil }
            return (config.cacheDir as NSString).appendingPathComponent("\(page).\(ext)")
        } else {
            guard page > 0 && page <= config.imageFiles.count else { return nil }
            
            let entry = config.imageFiles[page - 1]
            let ext = (entry.path as NSString).pathExtension
            guard !ext.isEmpty else { return nil }
            
            return (config.cacheDir as NSString).appendingPathComponent("\(page).\(ext)")
        }
    }

    /// Determine content type from filename
    private func getContentType(for filename: String) -> String {
        let ext = (filename as NSString).pathExtension.lowercased()
        switch ext {
        case "jpg", "jpeg":
            return "image/jpeg"
        case "png":
            return "image/png"
        case "gif":
            return "image/gif"
        case "webp":
            return "image/webp"
        case "bmp":
            return "image/bmp"
        case "svg":
            return "image/svg+xml"
        case "tiff", "tif":
            return "image/tiff"
        default:
            return "application/octet-stream"
        }
    }
}

File: apps/expo/modules/streamer/android/src/main/java/expo/modules/streamer/StreamerServer.kt
package expo.modules.streamer

import android.util.Log
import fi.iki.elonen.NanoHTTPD
import kotlinx.coroutines.*
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.ConcurrentHashMap

/**
 * Errors that can occur during streaming operations
 */
sealed class StreamerError : Exception() {
    data class ServerStartFailed(val reason: String) : StreamerError() {
        override val message: String get() = "Failed to start server: $reason"
    }

    data class ArchiveNotFound(val path: String) : StreamerError() {
        override val message: String get() = "Archive not found: $path"
    }

    data class ArchiveOpenFailed(val path: String, override val cause: Throwable) : StreamerError() {
        override val message: String get() = "Failed to open archive: $path - ${cause.message}"
    }

    data class PageNotFound(val page: Int) : StreamerError() {
        override val message: String get() = "Page $page not found in archive"
    }

    data class PageExtractionFailed(val page: Int, override val cause: Throwable) : StreamerError() {
        override val message: String get() = "Failed to extract page $page: ${cause.message}"
    }

    data class CacheDirectoryCreationFailed(override val cause: Throwable) : StreamerError() {
        override val message: String get() = "Failed to create cache directory: ${cause.message}"
    }
}

/**
 * Configuration for a book being streamed
 */
data class BookConfig(
    val bookId: String,
    val filePath: String,
    val cacheDir: String,
    // Cached list of image files in the archive (naturally sorted)
    var imageFiles: List<ZipArchive.Entry>? = null  
)

/**
 * Manages the HTTP server for streaming pages from ZIP/CBZ archives
 */
class StreamerServer private constructor() : NanoHTTPD("localhost", 0) {

    companion object {
        private const val TAG = "StreamerServer"
        val instance = StreamerServer()
    }

    /**
     * A cache of registered books (bookId -> BookConfig)
     */
    private val books = ConcurrentHashMap<String, BookConfig>()

    /**
     * Coroutine scope for background operations
     */
    private val extractionScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    /**
     * Whether the server is currently running
     */
    val isRunning: Boolean
        get() = isAlive

    /**
     * The port the server is running on
     */
    val serverPort: Int
        get() = listeningPort

    // MARK: - Server Lifecycle

    /**
     * Start the HTTP server if not already running
     * @return The actual port the server is running on
     * @throws StreamerError if server fails to start
     */
    fun startServer(): Int {
        if (isAlive) {
            return listeningPort
        }

        try {
            start(SOCKET_READ_TIMEOUT, false)
            Log.d(TAG, "Server started on port $listeningPort")
            return listeningPort
        } catch (e: Exception) {
            throw StreamerError.ServerStartFailed(e.message ?: "Unknown error")
        }
    }

    /**
     * Stop the HTTP server
     */
    fun stopServer() {
        stop()
        extractionScope.cancel()
        Log.d(TAG, "Server stopped")
    }

    // MARK: - Book Management

    /**
     * Register a book for streaming
     * @param config Book configuration
     * @throws StreamerError if archive not found or cache directory can't be created
     */
    fun registerBook(config: BookConfig) {
        val cleanFilePath = stripFilePrefix(config.filePath)
        val cleanCacheDir = stripFilePrefix(config.cacheDir)
        
        val file = File(cleanFilePath)
        if (!file.exists()) {
            throw StreamerError.ArchiveNotFound(cleanFilePath)
        }

        createCacheDirectoryIfNeeded(cleanCacheDir)

        val archive = try {
            ZipArchive(cleanFilePath)
        } catch (e: Exception) {
            throw StreamerError.ArchiveOpenFailed(cleanFilePath, e)
        }

        val imageFiles = archive.getImageFiles()
        archive.close()

        val updatedConfig = config.copy(
            filePath = cleanFilePath,
            cacheDir = cleanCacheDir,
            imageFiles = imageFiles
        )

        books[config.bookId] = updatedConfig
        Log.d(TAG, "Registered book ${config.bookId} with ${imageFiles.size} pages")
    }

    /**
     * Unregister a book and optionally clean up its cache
     * @param bookId The book ID to unregister
     * @param deleteCache Whether to delete cached pages
     */
    fun unregisterBook(bookId: String, deleteCache: Boolean = false) {
        val config = books.remove(bookId) ?: return

        if (deleteCache) {
            try {
                File(config.cacheDir).deleteRecursively()
                Log.d(TAG, "Deleted cache for book $bookId")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to delete cache for book $bookId", e)
            }
        }

        Log.d(TAG, "Unregistered book $bookId")
    }

    /**
     * Get the URL for a specific page
     * @param bookId The book ID
     * @param page The page number (1-indexed)
     * @return The URL to access the page, or null if server not running
     */
    fun getPageURL(bookId: String, page: Int): String? {
        if (!isAlive) return null
        return "http://localhost:$listeningPort/books/$bookId/pages/$page"
    }

    /**
     * Get the number of pages in a book
     * @param bookId The book ID
     * @return The number of pages, or 0 if book not found
     */
    fun getPageCount(bookId: String): Int {
        val config = books[bookId] ?: return 0
        return config.imageFiles?.size ?: 0
    }

    // MARK: - Request Handling

    override fun serve(session: IHTTPSession): Response {
        val uri = session.uri

        // Parse page request: /books/{bookId}/pages/{page}
        val pathParts = uri.split("/").filter { it.isNotEmpty() }

        if (pathParts.size != 4 || pathParts[0] != "books" || pathParts[2] != "pages") {
            return newFixedLengthResponse(Response.Status.BAD_REQUEST, "text/plain", "Invalid path")
        }

        val bookId = pathParts[1]
        val pageNumber = pathParts[3].toIntOrNull() ?: return newFixedLengthResponse(
            Response.Status.BAD_REQUEST,
            "text/plain",
            "Invalid page number"
        )

        val config = books[bookId] ?: return newFixedLengthResponse(
            Response.Status.NOT_FOUND,
            "text/plain",
            "Book not found: $bookId"
        )

        val cacheDir = File(config.cacheDir)
        val cachedFiles = cacheDir.listFiles { file ->
            file.nameWithoutExtension == pageNumber.toString()
        }

        if (cachedFiles != null && cachedFiles.isNotEmpty()) {
            return serveCachedPage(cachedFiles[0])
        }

        // Cache miss so extract
        return try {
            extractAndServePage(config, pageNumber)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to extract page $pageNumber from $bookId", e)
            newFixedLengthResponse(
                Response.Status.INTERNAL_ERROR,
                "text/plain",
                "Failed to extract page: ${e.message}"
            )
        }
    }

    // MARK: - Page Extraction

    /**
     * Extract a page from the archive and serve it
     */
    private fun extractAndServePage(config: BookConfig, page: Int): Response {
        val imageFiles = config.imageFiles ?: run {
            val archive = try {
                ZipArchive(config.filePath)
            } catch (e: Exception) {
                throw StreamerError.ArchiveOpenFailed(config.filePath, e)
            }
            
            val files = archive.getImageFiles()
            archive.close()
            files
        }

        if (page <= 0 || page > imageFiles.size) {
            throw StreamerError.PageNotFound(page)
        }

        // Note: page is 1-indexed
        val entry = imageFiles[page - 1]

        val archive = try {
            ZipArchive(config.filePath)
        } catch (e: Exception) {
            throw StreamerError.ArchiveOpenFailed(config.filePath, e)
        }

        val data = try {
            archive.extractEntry(entry)
        } catch (e: Exception) {
            archive.close()
            throw StreamerError.PageExtractionFailed(page, e)
        } finally {
            archive.close()
        }

        val extension = entry.name.substringAfterLast('.', "jpg")

        val cachedPath = getCachedPagePath(config, page, extension)
        try {
            FileOutputStream(cachedPath).use { output ->
                output.write(data)
            }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to cache page $page", e)
        }

        val contentType = getContentType(entry.name)
        Log.d(TAG, "Extracted and served page $page for book ${config.bookId}")

        return newFixedLengthResponse(Response.Status.OK, contentType, data.inputStream(), data.size.toLong())
    }

    /**
     * Serve a page from cache
     */
    private fun serveCachedPage(file: File): Response {
        val contentType = getContentType(file.name)
        Log.d(TAG, "Served cached page at ${file.path}")
        return newFixedLengthResponse(
            Response.Status.OK,
            contentType,
            file.inputStream(),
            file.length()
        )
    }

    // MARK: - Helper Methods

    /**
     * Create cache directory if it doesn't exist
     */
    private fun createCacheDirectoryIfNeeded(path: String) {
        val dir = File(path)
        if (!dir.exists()) {
            try {
                dir.mkdirs()
            } catch (e: Exception) {
                throw StreamerError.CacheDirectoryCreationFailed(e)
            }
        }
    }

    /**
     * Get the cached page path with proper extension
     */
    private fun getCachedPagePath(config: BookConfig, page: Int, extension: String = "jpg"): String {
        return File(config.cacheDir, "$page.$extension").absolutePath
    }

    /**
     * Strip file:// prefix from path if present
     */
    private fun stripFilePrefix(path: String): String {
        return if (path.startsWith("file://")) {
            path.substring(7)
        } else {
            path
        }
    }

    /**
     * Determine content type from filename
     */
    private fun getContentType(filename: String): String {
        return when (filename.substringAfterLast('.').lowercase()) {
            "jpg", "jpeg" -> "image/jpeg"
            "png" -> "image/png"
            "gif" -> "image/gif"
            "webp" -> "image/webp"
            else -> "application/octet-stream"
        }
    }
}
